
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../loculus/">
      
      
        <link rel="next" href="../schema/">
      
      
        
      
      
      <link rel="icon" href="../../assets/logo.svg">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.1">
    
    
      
        <title>Processing - sr2silo</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.484c7ddc.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.ab4e12ef.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="teal" data-md-color-accent="purple">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#processing-module" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="sr2silo" class="md-header__button md-logo" aria-label="sr2silo" data-md-component="logo">
      
  <img src="../../assets/logo.svg" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            sr2silo
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Processing
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="teal" data-md-color-accent="purple"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6m0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4M7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="teal" data-md-color-accent="lime"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/cbg-ethz/sr2silo" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    sr2silo
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../.." class="md-tabs__link">
        
  
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../usage/configuration/" class="md-tabs__link">
          
  
  
  Usage

        </a>
      </li>
    
  

      
        
  
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../loculus/" class="md-tabs__link">
          
  
  
  API Reference

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../contributing/" class="md-tabs__link">
          
  
  
  Contributing

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


  

<nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="sr2silo" class="md-nav__button md-logo" aria-label="sr2silo" data-md-component="logo">
      
  <img src="../../assets/logo.svg" alt="logo">

    </a>
    sr2silo
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/cbg-ethz/sr2silo" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    sr2silo
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Usage
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    Usage
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../usage/configuration/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Configuration
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../usage/organisms/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Multi-Organism Support
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../usage/deployment/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Deployment
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../usage/resource_requirements/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Resource Requirements
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" checked>
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    
  
    API Reference
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            
  
    API Reference
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../loculus/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Loculus Integration
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    
  
    Processing
  

    
  </span>
  
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    
  
    Processing
  

    
  </span>
  
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#data-structures" class="md-nav__link">
    <span class="md-ellipsis">
      
        Data Structures
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Data Structures">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#alignedread" class="md-nav__link">
    <span class="md-ellipsis">
      
        AlignedRead
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sr2silo.process.AlignedRead" class="md-nav__link">
    <span class="md-ellipsis">
      
        AlignedRead
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="AlignedRead">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#sr2silo.process.AlignedRead.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      
        __init__
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sr2silo.process.AlignedRead.__str__" class="md-nav__link">
    <span class="md-ellipsis">
      
        __str__
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sr2silo.process.AlignedRead.get_amino_acid_insertions" class="md-nav__link">
    <span class="md-ellipsis">
      
        get_amino_acid_insertions
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sr2silo.process.AlignedRead.get_metadata" class="md-nav__link">
    <span class="md-ellipsis">
      
        get_metadata
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sr2silo.process.AlignedRead.set_metadata" class="md-nav__link">
    <span class="md-ellipsis">
      
        set_metadata
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sr2silo.process.AlignedRead.set_nuc_insertion" class="md-nav__link">
    <span class="md-ellipsis">
      
        set_nuc_insertion
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sr2silo.process.AlignedRead.to_dict" class="md-nav__link">
    <span class="md-ellipsis">
      
        to_dict
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sr2silo.process.AlignedRead.to_silo_json" class="md-nav__link">
    <span class="md-ellipsis">
      
        to_silo_json
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gene" class="md-nav__link">
    <span class="md-ellipsis">
      
        Gene
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sr2silo.process.Gene" class="md-nav__link">
    <span class="md-ellipsis">
      
        Gene
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Gene">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#sr2silo.process.Gene.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      
        __init__
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sr2silo.process.Gene.to_dict" class="md-nav__link">
    <span class="md-ellipsis">
      
        to_dict
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#geneset" class="md-nav__link">
    <span class="md-ellipsis">
      
        GeneSet
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sr2silo.process.GeneSet" class="md-nav__link">
    <span class="md-ellipsis">
      
        GeneSet
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="GeneSet">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#sr2silo.process.GeneSet.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      
        __init__
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sr2silo.process.GeneSet.get_gene" class="md-nav__link">
    <span class="md-ellipsis">
      
        get_gene
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sr2silo.process.GeneSet.get_gene_length" class="md-nav__link">
    <span class="md-ellipsis">
      
        get_gene_length
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sr2silo.process.GeneSet.get_gene_name_list" class="md-nav__link">
    <span class="md-ellipsis">
      
        get_gene_name_list
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sr2silo.process.GeneSet.set_gene_length" class="md-nav__link">
    <span class="md-ellipsis">
      
        set_gene_length
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sr2silo.process.GeneSet.to_dict" class="md-nav__link">
    <span class="md-ellipsis">
      
        to_dict
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#nucinsertion" class="md-nav__link">
    <span class="md-ellipsis">
      
        NucInsertion
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sr2silo.process.NucInsertion" class="md-nav__link">
    <span class="md-ellipsis">
      
        NucInsertion
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="NucInsertion">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#aainsertion" class="md-nav__link">
    <span class="md-ellipsis">
      
        AAInsertion
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sr2silo.process.AAInsertion" class="md-nav__link">
    <span class="md-ellipsis">
      
        AAInsertion
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="AAInsertion">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#aainsertionset" class="md-nav__link">
    <span class="md-ellipsis">
      
        AAInsertionSet
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sr2silo.process.AAInsertionSet" class="md-nav__link">
    <span class="md-ellipsis">
      
        AAInsertionSet
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="AAInsertionSet">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#sr2silo.process.AAInsertionSet.__eq__" class="md-nav__link">
    <span class="md-ellipsis">
      
        __eq__
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sr2silo.process.AAInsertionSet.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      
        __init__
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sr2silo.process.AAInsertionSet.from_dict" class="md-nav__link">
    <span class="md-ellipsis">
      
        from_dict
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sr2silo.process.AAInsertionSet.set_insertions_for_gene" class="md-nav__link">
    <span class="md-ellipsis">
      
        set_insertions_for_gene
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sr2silo.process.AAInsertionSet.to_dict" class="md-nav__link">
    <span class="md-ellipsis">
      
        to_dict
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#bam-conversion" class="md-nav__link">
    <span class="md-ellipsis">
      
        BAM Conversion
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sr2silo.process.bam_to_sam" class="md-nav__link">
    <span class="md-ellipsis">
      
        bam_to_sam
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sr2silo.process.sam_to_bam" class="md-nav__link">
    <span class="md-ellipsis">
      
        sam_to_bam
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sr2silo.process.bam_to_fasta_query" class="md-nav__link">
    <span class="md-ellipsis">
      
        bam_to_fasta_query
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sr2silo.process.sort_bam_file" class="md-nav__link">
    <span class="md-ellipsis">
      
        sort_bam_file
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sr2silo.process.sort_and_index_bam" class="md-nav__link">
    <span class="md-ellipsis">
      
        sort_and_index_bam
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sr2silo.process.sort_sam_by_qname" class="md-nav__link">
    <span class="md-ellipsis">
      
        sort_sam_by_qname
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sr2silo.process.get_gene_set_from_ref" class="md-nav__link">
    <span class="md-ellipsis">
      
        get_gene_set_from_ref
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#translation-and-alignment" class="md-nav__link">
    <span class="md-ellipsis">
      
        Translation and Alignment
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sr2silo.process.nuc_to_aa_alignment" class="md-nav__link">
    <span class="md-ellipsis">
      
        nuc_to_aa_alignment
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sr2silo.process.parse_translate_align" class="md-nav__link">
    <span class="md-ellipsis">
      
        parse_translate_align
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sr2silo.process.parse_translate_align_in_batches" class="md-nav__link">
    <span class="md-ellipsis">
      
        parse_translate_align_in_batches
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sr2silo.process.curry_read_with_metadata" class="md-nav__link">
    <span class="md-ellipsis">
      
        curry_read_with_metadata
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#read-merging" class="md-nav__link">
    <span class="md-ellipsis">
      
        Read Merging
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sr2silo.process.paired_end_read_merger" class="md-nav__link">
    <span class="md-ellipsis">
      
        paired_end_read_merger
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#exceptions" class="md-nav__link">
    <span class="md-ellipsis">
      
        Exceptions
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sr2silo.process.ZeroFilteredReadsError" class="md-nav__link">
    <span class="md-ellipsis">
      
        ZeroFilteredReadsError
      
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../schema/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Data Schemas
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../vpipe/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    V-Pipe Integration
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Contributing
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            
  
    Contributing
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../contributing/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Overview
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../contributing/branching-strategy/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Branching Strategy
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="processing-module">Processing Module</h1>
<p>Core functions for BAM processing, translation, and alignment.</p>
<h2 id="data-structures">Data Structures</h2>
<h3 id="alignedread">AlignedRead</h3>


<div class="doc doc-object doc-class">



<h2 id="sr2silo.process.AlignedRead" class="doc doc-heading">
            <code>sr2silo.process.AlignedRead</code>


</h2>


    <div class="doc doc-contents first">



        <p>Class to represent an aligned read.</p>











  <div class="doc doc-children">










<div class="doc doc-object doc-function">


<h3 id="sr2silo.process.AlignedRead.__init__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">read_id</span><span class="p">,</span> <span class="n">unaligned_nucleotide_sequence</span><span class="p">,</span> <span class="n">aligned_nucleotide_sequence</span><span class="p">,</span> <span class="n">aligned_nucleotide_sequence_offset</span><span class="p">,</span> <span class="n">nucleotide_insertions</span><span class="p">,</span> <span class="n">amino_acid_insertions</span><span class="p">,</span> <span class="n">aligned_amino_acid_sequences</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Initialize with a AlignedRead object.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="sr2silo.process.AlignedRead.__str__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__str__</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>toString method as pretty JSON string.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="sr2silo.process.AlignedRead.get_amino_acid_insertions" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_amino_acid_insertions</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Return the amino acid insertions.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="sr2silo.process.AlignedRead.get_metadata" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_metadata</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Return the metadata.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="sr2silo.process.AlignedRead.set_metadata" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">set_metadata</span><span class="p">(</span><span class="n">metadata</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Set the metadata. Keep as ReadMetadata object to preserve alias information.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="sr2silo.process.AlignedRead.set_nuc_insertion" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">set_nuc_insertion</span><span class="p">(</span><span class="n">nuc_insertion</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Append a nucleotide insertion to the list of nucleotide insertions.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="sr2silo.process.AlignedRead.to_dict" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">to_dict</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Return a dictionary / json representation of the object.</p>
<p>Uses camelCase field names (via Pydantic aliases) for metadata fields
to match SILO database schema.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="sr2silo.process.AlignedRead.to_silo_json" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">to_silo_json</span><span class="p">(</span><span class="n">indent</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Validate the aligned read dict using the new SILO schema and return a
nicely formatted JSON string conforming to the DB requirements.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>indent</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to indent the JSON string, True for pretty print.</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>



  </div>

    </div>

</div><h3 id="gene">Gene</h3>


<div class="doc doc-object doc-class">



<h2 id="sr2silo.process.Gene" class="doc doc-heading">
            <code>sr2silo.process.Gene</code>


</h2>


    <div class="doc doc-contents first">



        <p>Class to represent a gene with a name and a length.</p>











  <div class="doc doc-children">










<div class="doc doc-object doc-function">


<h3 id="sr2silo.process.Gene.__init__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">gene_name</span><span class="p">,</span> <span class="n">gene_length</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Initialize with a gene name and a gene length.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="sr2silo.process.Gene.to_dict" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">to_dict</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Return a dictionary with gene name and gene length.</p>


    </div>

</div>



  </div>

    </div>

</div><h3 id="geneset">GeneSet</h3>


<div class="doc doc-object doc-class">



<h2 id="sr2silo.process.GeneSet" class="doc doc-heading">
            <code>sr2silo.process.GeneSet</code>


</h2>


    <div class="doc doc-contents first">



        <p>Class to represent a set of genes for a pathogen</p>











  <div class="doc doc-children">










<div class="doc doc-object doc-function">


<h3 id="sr2silo.process.GeneSet.__init__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">genes</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Initialize with a list of genes.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="sr2silo.process.GeneSet.get_gene" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_gene</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Return a gene by name.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="sr2silo.process.GeneSet.get_gene_length" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_gene_length</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Return the length of a gene.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="sr2silo.process.GeneSet.get_gene_name_list" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_gene_name_list</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Return a list of genes.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="sr2silo.process.GeneSet.set_gene_length" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">set_gene_length</span><span class="p">(</span><span class="n">gene_name</span><span class="p">,</span> <span class="n">gene_length</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Set the length of a gene.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="sr2silo.process.GeneSet.to_dict" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">to_dict</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Return a dictionary with gene names as keys and gene
length as values.</p>


    </div>

</div>



  </div>

    </div>

</div><h3 id="nucinsertion">NucInsertion</h3>


<div class="doc doc-object doc-class">



<h2 id="sr2silo.process.NucInsertion" class="doc doc-heading">
            <code>sr2silo.process.NucInsertion</code>


</h2>


    <div class="doc doc-contents first">
            <p class="doc doc-class-bases">
              Bases: <code><span title="sr2silo.process.interface.Insertion">Insertion</span></code></p>



        <p>A nuclotide insertion.</p>











  <div class="doc doc-children">












  </div>

    </div>

</div><h3 id="aainsertion">AAInsertion</h3>


<div class="doc doc-object doc-class">



<h2 id="sr2silo.process.AAInsertion" class="doc doc-heading">
            <code>sr2silo.process.AAInsertion</code>


</h2>


    <div class="doc doc-contents first">
            <p class="doc doc-class-bases">
              Bases: <code><span title="sr2silo.process.interface.Insertion">Insertion</span></code></p>



        <p>An amino acid insertion.</p>











  <div class="doc doc-children">












  </div>

    </div>

</div><h3 id="aainsertionset">AAInsertionSet</h3>


<div class="doc doc-object doc-class">



<h2 id="sr2silo.process.AAInsertionSet" class="doc doc-heading">
            <code>sr2silo.process.AAInsertionSet</code>


</h2>


    <div class="doc doc-contents first">



        <p>Class to represent the set of amino acid insertions for a full set of genes.</p>











  <div class="doc doc-children">










<div class="doc doc-object doc-function">


<h3 id="sr2silo.process.AAInsertionSet.__eq__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__eq__</span><span class="p">(</span><span class="n">other</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Compare two AAInsertionSet objects by their values.</p>
<p>This compares the dictionaries generated by to_dict() method,
ensuring a value-based comparison rather than identity-based.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="sr2silo.process.AAInsertionSet.__init__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">genes</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Initialize with an empty set of insertions for each gene.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="sr2silo.process.AAInsertionSet.from_dict" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">from_dict</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-staticmethod"><code>staticmethod</code></small>
  </span>

</h3>


    <div class="doc doc-contents ">

        <p>Create an AAInsertionSet object from a dictionary.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="sr2silo.process.AAInsertionSet.set_insertions_for_gene" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">set_insertions_for_gene</span><span class="p">(</span><span class="n">gene_name</span><span class="p">,</span> <span class="n">aa_insertions</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Set the amino acid insertions for a particular gene.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="sr2silo.process.AAInsertionSet.to_dict" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">to_dict</span><span class="p">()</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Return a dictionary with gene names as keys.</p>


    </div>

</div>



  </div>

    </div>

</div><h2 id="bam-conversion">BAM Conversion</h2>


<div class="doc doc-object doc-function">


<h2 id="sr2silo.process.bam_to_sam" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">sr2silo</span><span class="o">.</span><span class="n">process</span><span class="o">.</span><span class="n">bam_to_sam</span><span class="p">(</span><span class="n">bam_file</span><span class="p">,</span> <span class="n">sam_file</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents first">

        <p>Converts a BAM file to SAM format and writes it to the specified output file.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>bam_file</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the input BAM file.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>sam_file</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the output SAM file.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>sam_file</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the output SAM file.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="sr2silo.process.sam_to_bam" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">sr2silo</span><span class="o">.</span><span class="n">process</span><span class="o">.</span><span class="n">sam_to_bam</span><span class="p">(</span><span class="n">sam_file</span><span class="p">,</span> <span class="n">bam_file</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents first">

        <p>Convert a SAM file to a BAM file.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>sam_file</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the input SAM file.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>bam_file</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the output BAM file.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="sr2silo.process.bam_to_fasta_query" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">sr2silo</span><span class="o">.</span><span class="n">process</span><span class="o">.</span><span class="n">bam_to_fasta_query</span><span class="p">(</span><span class="n">bam_file</span><span class="p">,</span> <span class="n">fasta_file</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents first">

        <p>Convert a BAM file to a FASTA file. Bluntly resolved the sam to fasta,
removing soft clippings, keeping insertions and deletions, ignoring skipped
regions and paddings.</p>
<p>Outputs the sequence as it is read from the molecule.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>bam_file</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the input BAM file.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>fasta_file</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the output FASTQ file.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="sr2silo.process.sort_bam_file" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">sr2silo</span><span class="o">.</span><span class="n">process</span><span class="o">.</span><span class="n">sort_bam_file</span><span class="p">(</span><span class="n">input_bam_path</span><span class="p">,</span> <span class="n">output_bam_path</span><span class="p">,</span> <span class="n">sort_by_qname</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents first">

        <p>Sorts a BAM file using pysam.sort by alignment positions by default,
but can also sort by query name if specified.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>input_bam_path</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the input BAM file.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>output_bam_path</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the output sorted BAM file.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>sort_by_qname</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, sorts by query name. Defaults to False.</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="sr2silo.process.sort_and_index_bam" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">sr2silo</span><span class="o">.</span><span class="n">process</span><span class="o">.</span><span class="n">sort_and_index_bam</span><span class="p">(</span><span class="n">input_bam_fp</span><span class="p">,</span> <span class="n">output_bam_fp</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents first">

        <p>Function to sort and index the input BAM file,</p>
<p>implements checks to see if the input BAM file is already sorted and indexed.</p>
<p>If not sorted and indexed, the function will sort and index the input BAM file.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="sr2silo.process.sort_sam_by_qname" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">sr2silo</span><span class="o">.</span><span class="n">process</span><span class="o">.</span><span class="n">sort_sam_by_qname</span><span class="p">(</span><span class="n">input_sam_path</span><span class="p">,</span> <span class="n">output_sam_path</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents first">

        <p>Sorts a sam file using pysam.sort command by query name.
Args:
    input_bam_path (Path): Path to the input BAM file.
    output_bam_path (Path): Path to the output sorted BAM file.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="sr2silo.process.get_gene_set_from_ref" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">sr2silo</span><span class="o">.</span><span class="n">process</span><span class="o">.</span><span class="n">get_gene_set_from_ref</span><span class="p">(</span><span class="n">reference_fp</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents first">

        <p>Load the gene ref fasta and create a GeneSet with gene short
names and lengths.</p>


    </div>

</div><h2 id="translation-and-alignment">Translation and Alignment</h2>


<div class="doc doc-object doc-function">


<h2 id="sr2silo.process.nuc_to_aa_alignment" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">sr2silo</span><span class="o">.</span><span class="n">process</span><span class="o">.</span><span class="n">nuc_to_aa_alignment</span><span class="p">(</span><span class="n">in_nuc_alignment_fp</span><span class="p">,</span> <span class="n">in_aa_db_fp</span><span class="p">,</span> <span class="n">out_aa_alignment_fp</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents first">

        <p>Function to convert files and translate and align with Diamond / blastx.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>in_nuc_alignment_fp</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the input nucleotide alignment file.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>in_aa_db_fp</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the input amino acid diamond database file.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>out_aa_alignment_fp</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the output amino acid alignment file.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>None</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>None</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="description" open>
  <summary>Description</summary>
  <p>Uses Diamond with the settings:
--evalue 1
--gapopen 6
--gapextend 2
--outfmt 101
--matrix BLOSUM62
--unal 0
--max-hsps 1
--block-size 0.5</p>
</details>

    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="sr2silo.process.parse_translate_align" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">sr2silo</span><span class="o">.</span><span class="n">process</span><span class="o">.</span><span class="n">parse_translate_align</span><span class="p">(</span><span class="n">nuc_reference_fp</span><span class="p">,</span> <span class="n">aa_reference_fp</span><span class="p">,</span> <span class="n">nuc_alignment_fp</span><span class="p">,</span> <span class="n">aa_db_fp</span><span class="p">,</span> <span class="n">target_reference</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents first">

        <p>Parse nucleotides, translate and align amino acids the input files.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>nuc_reference_fp</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to nucleotide reference FASTA file.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>aa_reference_fp</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to amino acid reference FASTA file.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>nuc_alignment_fp</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to nucleotide alignment BAM file.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>aa_db_fp</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to Diamond database file.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>target_reference</code>
            </td>
            <td>
                  <code><span title="str">str</span> | None</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Filter reads to only include those aligned to this
             reference accession. If None, all reads are processed.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.Dict">Dict</span>[<span title="str">str</span>, <a class="autorefs autorefs-internal" title="sr2silo.process.AlignedRead (sr2silo.process.interface.AlignedRead)" href="#sr2silo.process.AlignedRead">AlignedRead</a>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary mapping read IDs to AlignedRead objects.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="sr2silo.process.parse_translate_align_in_batches" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">sr2silo</span><span class="o">.</span><span class="n">process</span><span class="o">.</span><span class="n">parse_translate_align_in_batches</span><span class="p">(</span><span class="n">nuc_reference_fp</span><span class="p">,</span> <span class="n">aa_reference_fp</span><span class="p">,</span> <span class="n">nuc_alignment_fp</span><span class="p">,</span> <span class="n">metadata_fp</span><span class="p">,</span> <span class="n">output_fp</span><span class="p">,</span> <span class="n">chunk_size</span><span class="o">=</span><span class="mi">100000</span><span class="p">,</span> <span class="n">write_chunk_size</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">target_reference</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents first">

        <p>Parse nucleotides, translate and align amino acids in batches.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>nuc_reference_fp</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the nucleotide reference genome - .fasta</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>aa_reference_fp</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the amino acid reference genome - .fasta</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>nuc_alignment_fp</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the nucleotide alignment file - .bam</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>metadata_fp</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the metadata file - .json</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>output_fp</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the output file - .ndjson</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>chunk_size</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Size of each batch, in number of reads.</p>
              </div>
            </td>
            <td>
                  <code>100000</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>write_chunk_size</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Size of each write batch.</p>
              </div>
            </td>
            <td>
                  <code>20</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>target_reference</code>
            </td>
            <td>
                  <code><span title="str">str</span> | None</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Filter reads to only include those aligned
to this reference accession. Should match @SQ SN field in BAM header.
If None, all reads are processed.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>Path</code></td>            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The path to the output file with the correct suffix.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="resources" open>
  <summary>Resources</summary>
  <p>A chunk_size of 100000 reads is a good starting point for most cases.
This will take about 3.7 GB ram for Covid Genomes and 1-2 minutes to process.</p>
</details>

<details class="constraints" open>
  <summary>Constraints</summary>
  <p>No sorting constraints are enforced on the input files.
The output file is compressed with zstd.</p>
</details>

<details class="constraints" open>
  <summary>Constraints</summary>
  <p>No sorting constraints are enforced on the input files.
The output file is compressed with zstd.</p>
</details>

<details class="logs" open>
  <summary>Logs</summary>
  <p>All logs of INFO and below are suppressed.</p>
</details>

    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="sr2silo.process.curry_read_with_metadata" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">sr2silo</span><span class="o">.</span><span class="n">process</span><span class="o">.</span><span class="n">curry_read_with_metadata</span><span class="p">(</span><span class="n">metadata_fp</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents first">

        <p>Returns a function that enriches an AlignedRead with metadata.</p>
<p>Reads metadata from a JSON file and returns a function that enriches
an AlignedRead with the loaded metadata.</p>


    </div>

</div><h2 id="read-merging">Read Merging</h2>


<div class="doc doc-object doc-function">


<h2 id="sr2silo.process.paired_end_read_merger" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">sr2silo</span><span class="o">.</span><span class="n">process</span><span class="o">.</span><span class="n">paired_end_read_merger</span><span class="p">(</span><span class="n">nuc_align_sam_fp</span><span class="p">,</span> <span class="n">ref_genome_fasta_fp</span><span class="p">,</span> <span class="n">output_merged_sam_fp</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents first">

        <p>Merge paired-end reads using <code>smallgenomutilities</code>tool <code>paired_end_read_merger</code>.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>nuc_align_sam_fp</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the nucleotide alignment SAM file.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>ref_genome_fasta_fp</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the reference genome FASTA file.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>output_merged_sam_fp</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the output merged SAM file.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>None</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>None</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    </div>

</div><h2 id="exceptions">Exceptions</h2>


<div class="doc doc-object doc-class">



<h2 id="sr2silo.process.ZeroFilteredReadsError" class="doc doc-heading">
            <code>sr2silo.process.ZeroFilteredReadsError</code>


</h2>


    <div class="doc doc-contents first">
            <p class="doc doc-class-bases">
              Bases: <code><span title="Exception">Exception</span></code></p>



        <p>Raised when reference filtering results in zero reads.</p>
<p>This error indicates that a target reference was specified for filtering,
but no reads in the BAM file aligned to that reference. This typically
means the wrong reference was specified or the data doesn't contain
reads for the expected reference.</p>









    </div>

</div>












                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        
<div class="md-social">
  
    
    
    
    
      
      
    
    <a href="https://github.com/cbg-ethz/sr2silo" target="_blank" rel="noopener" title="github.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M202.1 328.7c0 20.9-10.9 55.1-36.7 55.1s-36.7-34.2-36.7-55.1 10.9-55.1 36.7-55.1 36.7 34.2 36.7 55.1M496 278.2c0 31.9-3.2 65.7-17.5 95-37.9 76.6-142.1 74.8-216.7 74.8-75.8 0-186.2 2.7-225.6-74.8-14.6-29-20.2-63.1-20.2-95 0-41.9 13.9-81.5 41.5-113.6-5.2-15.8-7.7-32.4-7.7-48.8 0-21.5 4.9-32.3 14.6-51.8 45.3 0 74.3 9 108.8 36 29-6.9 58.8-10 88.7-10 27 0 54.2 2.9 80.4 9.2 34-26.7 63-35.2 107.8-35.2 9.8 19.5 14.6 30.3 14.6 51.8 0 16.4-2.6 32.7-7.7 48.2 27.5 32.4 39 72.3 39 114.2m-64.3 50.5c0-43.9-26.7-82.6-73.5-82.6-18.9 0-37 3.4-56 6-14.9 2.3-29.8 3.2-45.1 3.2-15.2 0-30.1-.9-45.1-3.2-18.7-2.6-37-6-56-6-46.8 0-73.5 38.7-73.5 82.6 0 87.8 80.4 101.3 150.4 101.3h48.2c70.3 0 150.6-13.4 150.6-101.3m-82.6-55.1c-25.8 0-36.7 34.2-36.7 55.1s10.9 55.1 36.7 55.1 36.7-34.2 36.7-55.1-10.9-55.1-36.7-55.1"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": "../..", "features": ["navigation.tabs", "navigation.sections", "toc.integrate", "search.suggest", "search.highlight", "content.tabs.link", "content.code.annotation", "content.code.copy"], "search": "../../assets/javascripts/workers/search.2c215733.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../assets/javascripts/bundle.79ae519e.min.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  </body>
</html>